### PR TITLE
ISLANDORA-2141 Don't double-encode link to object.

### PR DESCRIPTION
**Pull request by rosiel**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2141
Actually re-fixes the problem from https://jira.duraspace.org/browse/ISLANDORA-1994

# What does this Pull Request do?

Changes how we solve the problem of parse_url failing on pids above namespace:65535, because the previous fix (urlencode()) caused unintended consequences. The final link was double-encoded (namespace%253A65536) and while Drupal still rendered the right object, other stuff broke:
- Usage Stats views (that appear on the collection management page) no longer showed up
- contextual views (which rely on the PID from the URL) no longer worked. 

# What's new?

Don't double-encode; add a meaningless fragment to the url instead. (this tricks parse_url, which is the thing that's really at fault...) 

# How should this be tested?

Create an object with a PID of namespace:65536 or greater. (this is easily done by creating a collection; the collection form lets you assign a PID)
Create a view, with a field that "links to object" and shows this object.

Before the PR: the link renders, and includes %253A in place of colon (:) 
After the PR: the link renders, and includes #view at the end.

To test with Usage Stats
- enable usage stats and make sure you have an object with some stats available.
- get to that object from a "link to object" field in a solr view
Before this PR: the collection usage stats table doesn't show up
After this PR: the collection usage stats table shows up.

# Additional Notes:

* Does this change require documentation to be updated? **Nope**
* Does this change add any new dependencies? **Nope**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)?  **Nope**
* Could this change impact execution of existing code? **Maybe!? Hopefully less than the previous solution?**

# Interested parties
@Islandora/7-x-1-x-committers and @bondjimbond

### CODE REVIEW

**Comment by DiegoPino** [comment]
Hi @rosiel , could you please add to the description that this only affects incremental PIDS (numerical part) from 65536 to 99999? 
Also, would a simple `#` at the end without the`view` piece work too? I rather prefer links not jumping to a `view` anchor if that happens to exist!. Wonder if we could ask @giancarlobi to test/confirm this? He uses a lot of Views in his system and has probably been affected by this somewhere/somewhen. Thanks

**Comment by bondjimbond** [comment]
@DiegoPino This doesn't only affect high-numerical PIDs, it affects all PIDs. The earlier pull that caused this problem was created to resolve an issue with pids 65535 and above; its unintended consequences affect every object.

**Comment by DiegoPino** [comment]
@bondjimbond you got me wrong there. The need for a `#view`is only for numerical PIDs between  65536 to 99999. The issue right now with double encoding, can be for all of them, i have not tested it because i don't use usage stats at all and other uses seem to be unaffected? (like the view used to display ingested objects? maybe the link there is fixed and not build using this.. who knows).
But again: double encoding bad, but adding hash really only needed for numerical PIDs between  65536 to 99999. Does that sound better?

**Comment by rosiel** [comment]
@DiegoPino sure - a more targeted approach - just on the pids that were affected by the first issue - would work.
To answer the earlier question, no, `#` without a fragment string after it does not work - parse_url() still fails. Would you rather a different fragment string? I'm open to suggestions. #view seemed non-intrusive but could, potentially, be used on the page as an intentional anchor.

**Comment by rosiel** [comment]
@DiegoPino does this latest commit address your concerns?

**Comment by DiegoPino** [comment]
@rosiel yes, looks legit to me. Thanks. And that parsing should pretty fast right? But what i specially like that since the # thing is a trick that works for URL parsing (good, never new about that) then `#_`seems the safest way. thanks for that

**Comment by rosiel** [comment]
I avoided regex by using explode(), so it's as fast as I know how to make it (speed-ups welcome!) 

I figured out the # thing when I tried to "Add link" to a Drupal menu. A link path like `islandora/object/islandora:123` will *not* work, but `islandora/object/islandora:123#` will. (It forgets about the # once you've saved the link, oddly enough...) 

:)

**Review by DiegoPino — APPROVED** [review]
Works. Less intrusive than previous solution

**Comment by DiegoPino** [comment]
Cool. Wonder If Drupal in its internals  forgets about `the #`a.k.a fragment then maybe there is even something more we can use to make sure we don´t double encode ... maybe out render array is weird? not sure i can tackle on that, but your solution works good for me. Will approve and wait for 24 and ask also @MorganDawe for feedback since previous fix came from one of their users needs.

**Comment by DiegoPino** [comment]
@MorganDawe could i get your comments here? This pull is approved but wanted your reactions before merging since previous fix came from you/your user's needs. Thanks!

**Comment by bondjimbond** [comment]
@DiegoPino No response to your request for comment 19 days ago... are we safe to merge this?

**Comment by MorganDawe** [comment]
@DiegoPino my most sincere apologies for not responding sooner. This looks like it will absolutely address client support requirements, avoiding treating pids as port numbers. Thank you @rosiel , @bondjimbond and @DiegoPino for your work resolving this issue.

**Pull request merged**
